src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs

```diff
@@ -0,0 +1,243 @@
+        // LinkedList is thread safe for Add, Remove, and Clear operations.
+        private static LinkedList<MeterListener> s_allStartedListeners = new LinkedList<MeterListener>();
+
+        // List of the instruments which the current listener is listening to.
+        private LinkedList<Instrument> _enabledMeasurementInstruments = new LinkedList<Instrument>();
+        private bool _disposed;
+
+        // We initialize all measurement callback with no-op operations so we'll avoid the null checks during the execution;
+        private MeasurementCallback<byte>    _byteMeasurementCallback    = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<short>   _shortMeasurementCallback   = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<int>     _intMeasurementCallback     = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<long>    _longMeasurementCallback    = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<float>   _floatMeasurementCallback   = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<double>  _doubleMeasurementCallback  = (instrument, measurement, tags, state) => { /* no-op */ };
+        private MeasurementCallback<decimal> _decimalMeasurementCallback = (instrument, measurement, tags, state) => { /* no-op */ };
+
+        /// <summary>
+        /// Creates a MeterListener object.
+        /// </summary>
+        public MeterListener() { }
+
+        /// <summary>
+        /// Callbacks to get notification when an instrument is published.
+        /// </summary>
+        public Action<Instrument, MeterListener>? InstrumentPublished { get; set; }
+
+        /// <summary>
+        /// Callbacks to get notification when stopping the measurement on some instrument.
+        /// This can happen when the Meter or the Listener is disposed or calling <see cref="Dispose" /> on the listener.
+        /// </summary>
+        public Action<Instrument, object?>? MeasurementsCompleted { get; set; }
+
+        /// <summary>
+        /// Start listening to a specific instrument measurement recording.
+        /// </summary>
+        /// <param name="instrument">The instrument to listen to.</param>
+        /// <param name="state">A state object which will be passed back to the callback getting measurements events.</param>
+        public void EnableMeasurementEvents(Instrument instrument, object? state = null)
+        {
+            if (instrument is null || _disposed)
+            {
+                return;
+            }
+
+            _enabledMeasurementInstruments.AddIfNotExist(instrument, (instrument1, instrument2) => object.ReferenceEquals(instrument1, instrument2));
```

Comment:
In the prototype code, this:
```C#
listener.EnableMeasurementEvents(instrument, cookie1);
listener.EnableMeasurementEvents(instrument, cookie2);
```
Is equivalent to:
```C#
listener.EnableMeasurementEvents(instrument, cookie1);
listener.DisableMeasurementEvents(instrument);
listener.EnableMeasurementEvents(instrument, cookie2);
```

Here the behavior appears to ignore the new cookie and leave the instrument enabled with cookie1. IMO the prototype's behavior is more intuitive than this one but I am happy to hear if others have different opinions.

Reply:
I'll go with the prototype behavior as I liked more too. I'll add tests too.